packages/block-editor/src/utils/dom.js

```diff
@@ -57,3 +57,98 @@ export function getBlockClientId( node ) {
 
 	return blockNode.id.slice( 'block-'.length );
 }
+
+/**
+ * Returns the union of two DOMRect objects.
+ *
+ * @param {DOMRect} rect1 First rectangle.
+ * @param {DOMRect} rect2 Second rectangle.
+ * @return {DOMRect} Union of the two rectangles.
+ */
+export function rectUnion( rect1, rect2 ) {
+	const left = Math.min( rect1.left, rect2.left );
+	const top = Math.min( rect1.top, rect2.top );
```

Comment:
> On thing I noticed is that the toolbar align itself to the right - from what I can tell it's the inline translateX value of the block-editor-block-list__block-popover element.

Could be due to negative values being misinterpreted by the Popover component.

If I pass `0` in the case of negative values — `left = left < 0 ? 0 : left` — things look okay 🤔 

https://github.com/user-attachments/assets/203cadab-40be-44e7-9a26-ac7859114c5b

I'm not 100% on this. Smells like a Popover thing.

Reply:
Hmm. Can you share the markup you have in the template?

Reply:
Nothing out of the ordinary I think. Here's the template:

```html
<!-- wp:template-part {"slug":"header","theme":"emptytheme","tagName":"header"} /-->

<!-- wp:query {"queryId":1,"query":{"offset":0,"postType":"post","order":"desc","orderBy":"date","author":"","search":"","sticky":"","perPage":10}} -->
<div class="wp-block-query"><!-- wp:post-template -->
<!-- wp:post-title {"isLink":true} /-->

<!-- wp:post-excerpt /-->
<!-- /wp:post-template --></div>
<!-- /wp:query -->
```

And the header template part:

```html
<!-- wp:navigation {"ref":6} /-->

<!-- wp:site-title /-->

<!-- wp:site-tagline /-->
```

Does that help? 😄

Reply:
Hmm there is probably some hidden element that makes `getVisibleElementBounds` return a bounds that's larger than it should be, but I'm having a hard time reproducing. Could you export your theme? Or come up with steps to reproduce that begin with installing Emptytheme?

Reply:
Try the latest commit I just pushed @ramonjd. I updated our check for hidden elements to explicitly check for `<VisuallyHidden>` which I think is causing problems.

Reply:
>  Could you export your theme? Or come up with steps to reproduce that begin with installing Emptytheme?

Hmm, it's still happening for me on Firefox, Chrome and Safari. 

It's just empty theme on a clean install. 

Head to the site editor and add a navigation block as the first element in the header.

https://github.com/user-attachments/assets/d5ddc1f0-c0e0-48b7-8a81-cb82d87d2089

Reply:
Following Ramon's steps above I can also replicate the misaligned toolbar in Chrome, Firefox, Safari etc.


https://github.com/user-attachments/assets/ce39d2c0-bec8-4cf2-a059-34dace295446

Reply:
Thanks @aaronrobertshaw 

If I get time, I might play around with disallowing negative values. I'm not super confident about that approach but hacking never hurt anyone, except maybe some major corporations, millions of consumers and certain small democracies.

Reply:
I'm not sure where the calculations are happening, but a `console.trace` confirms that that the `-1` value is running through these new functions

```
{left: -1, top: 0, width: 1781, height: 63}
rectUnion	@	dom.js:73
getVisibleElementBounds	@	dom.js:142
getBoundingClientRect	@	index.js:94
```


But after looking in the wrong place for ages, from what I can tell, it's because the selected submenu has a left value of `-1px` in its editor styles

https://github.com/WordPress/gutenberg/blob/ae2d23d890614d5d538b3fe25b9a12163d198687/packages/block-library/src/navigation-submenu/editor.scss#L20-L20

https://github.com/WordPress/gutenberg/blob/ae2d23d890614d5d538b3fe25b9a12163d198687/packages/block-library/src/navigation/style.scss#L197-L197

Remove these, and it's kosher.

https://github.com/user-attachments/assets/4f653b7a-0e01-476c-a481-3dd0ac208016

So, I suppose there are two choices as far as I can tell:

1. Remove the -1px rules on the submenu (the quickest fix). I tested [the bug it was trying to fix](https://github.com/WordPress/gutenberg/pull/35716) and I can't reproduce with the values removed, but it doesn't address the primary cause.
2."Normalize" negative values in `rectUnion` (e.g.,round up to zero in the case of `x`) to calculate the **[visible element bounds](https://github.com/WordPress/gutenberg/blob/ae2d23d890614d5d538b3fe25b9a12163d198687/packages/block-editor/src/utils/dom.js#L142-L142)**. On the surface this makes the most sense since we're only interested in what's visible in the viewport.

Reply:
> "Normalize" negative values in rectUnion (e.g.,round up to zero in the case of x) to calculate the visible element bounds. On the surface this makes the most sense since we're only interested in what's visible in the viewport.

<details>

<summary>This is what I was thinking</summary>

```diff
diff --git a/packages/block-editor/src/utils/dom.js b/packages/block-editor/src/utils/dom.js
index eef322959a..2504c7824b 100644
--- a/packages/block-editor/src/utils/dom.js
+++ b/packages/block-editor/src/utils/dom.js
@@ -65,12 +65,25 @@ export function getBlockClientId( node ) {
  * @param {DOMRect} rect2 Second rectangle.
  * @return {DOMRect} Union of the two rectangles.
  */
-export function rectUnion( rect1, rect2 ) {
-	const left = Math.min( rect1.left, rect2.left );
-	const top = Math.min( rect1.top, rect2.top );
-	const right = Math.max( rect1.right, rect2.right );
-	const bottom = Math.max( rect1.bottom, rect2.bottom );
-	return new window.DOMRect( left, top, right - left, bottom - top );
+export function rectUnion( rect1, rect2, viewportRect ) {
+	let left = Math.min( rect1.left, rect2.left );
+	let top = Math.min( rect1.top, rect2.top );
+	let right = Math.max( rect1.right, rect2.right );
+	let bottom = Math.max( rect1.bottom, rect2.bottom );
+
+	if ( viewportRect ) {
+		left = Math.max( left, viewportRect.left );
+		top = Math.max( top, viewportRect.top );
+		right = Math.min( right, viewportRect.right );
+		bottom = Math.min( bottom, viewportRect.bottom );
+	}
+
+	return new window.DOMRect(
+		left,
+		top,
+		right - left,
+		bottom - top
+	);
 }
 
 /**
@@ -130,7 +143,12 @@ export function getVisibleElementBounds( element ) {
 	}
 
 	let bounds = element.getBoundingClientRect();
-
+	const viewportRect = {
+		top: 0,
+		right: viewport.innerWidth,
+		bottom: viewport.innerHeight,
+		left: 0,
+	};
 	const stack = [ element ];
 	let currentElement;
 
@@ -138,7 +156,7 @@ export function getVisibleElementBounds( element ) {
 		for ( const child of currentElement.children ) {
 			if ( isElementVisible( child ) ) {
 				const childBounds = child.getBoundingClientRect();
-				bounds = rectUnion( bounds, childBounds );
+				bounds = rectUnion( bounds, childBounds, viewportRect );
 				stack.push( child );
 			}
 		}

```

</details>

Thoughts?

Reply:
> Remove the -1px rules on the submenu

Without looking, I assume this is just to prevent the appearance of a double border along the window edge. That sounds like something we'd visually want to keep.

> "Normalize" negative values in rectUnion

In the absence of the first option. This seems like the winner by default 🤷 

> Thoughts?

Looks viable to me. There's a lot of `Math.min/max` going on, would a quick inline comment help our future selves grok why the viewport rect is used etc?

Reply:
> would a quick inline comment help our future selves grok why the viewport rect is used etc?

Definitely. Thanks for the sanity check. I'll commit this change. Appears to work as intended.